File: history/javascript-v0.3.0/tools/deriveprofiles.js
// This Source Code Form is subject to the terms of the Mozilla Public
// License, v. 2.0. If a copy of the MPL was not distributed with this
// file, You can obtain one at https://mozilla.org/MPL/2.0/.

// Derives the donor profile table of spec section 4.2.
//
// A profile is an ordered ranking of seven alphabet characters. A segment
// whose mask has k bits set spends the first k of them as stand-ins, so only
// those k become unusable as literals inside it -- which is why the table is
// searched by encoded size and not by character frequency: what matters is
// how often a donor's own character turns up in the text it would have to
// stand in for.
//
// The search is greedy in two nested ways. Profiles are added one at a time,
// each one chosen to help most given the ones already in the table; within a
// profile, donor positions are filled left to right. Position r only has an
// effect on segments with k > r, and k > 3 is rare, so the later positions
// mostly fall back to the rarity order -- which is the right default anyway.
//
// Usage:  node tools/deriveprofiles.js [count]

import { readFileSync, readdirSync } from 'node:fs';
import { join } from 'node:path';
import { makeCodec, ALPHABET, R_LEN } from '../src/codec.js';
import { BENCH_DIR } from '../bench/lib.js';

// Like CORPUS_DIR, the training data stays at the repository root: it is
// written by tools/traincorpus.py, which derives a table the current
// specification still carries.
const TRAIN_DIR = join(BENCH_DIR, '..', '..', '..', 'bench', 'train');
const WANT = Number(process.argv[2] ?? 32);
const POOL_SIZE = 20;

const files = readdirSync(TRAIN_DIR)
  .filter((n) => n.endsWith('.train'))
  .sort()
  .map((n) => new Uint8Array(readFileSync(join(TRAIN_DIR, n))));
if (!files.length) {
  throw new Error(`no training data -- run: python3 tools/traincorpus.py`);
}
const totalIn = files.reduce((s, f) => s + f.length, 0);
process.stderr.write(`training on ${files.length} files, ${totalIn} bytes\n`);

// Candidate pool: the rarest alphabet characters in the training text.
const counts = new Map([...ALPHABET].map((c) => [c, 0]));
for (const f of files) {
  for (const b of f) {
    const ch = String.fromCharCode(b);
    if (counts.has(ch)) counts.set(ch, counts.get(ch) + 1);
  }
}
// Letters and digits are excluded from the pool on principle rather than by
// frequency. A rare capital is rare when all text is counted together and
// common in the one file that happens to use it -- identifiers and words are
// made of letters -- so a letter donor breaks segments in bursts. Measured,
// the difference runs the wrong way on the training corpus and the right way
// on the hold-out: letters win by 0.1 % where they were derived and lose by
// 0.1 % where they were not (bench/results/RESULTS.md).
const pool = [...counts.entries()]
  .filter(([c]) => c !== '-' && !/[0-9A-Za-z]/.test(c))
  .sort((a, b) => a[1] - b[1])
  .slice(0, POOL_SIZE)
  .map(([c]) => c);
process.stderr.write(`pool: ${pool.join(' ')}\n`);

const cost = (profiles) => {
  const codec = makeCodec({ profiles, minDpBytes: 24, minBinaryRun: 1 });
  let n = 0;
  for (const f of files) n += codec.encode(f).length;
  return n;
};

const rarityFill = (prefix) => {
  const out = [...prefix];
  for (const c of pool) {
    if (out.length === R_LEN) break;
    if (!out.includes(c)) out.push(c);
  }
  return out;
};

const table = [];
const curve = [];
for (let p = 0; p < WANT; p++) {
  const chosen = [];
  for (let r = 0; r < R_LEN; r++) {
    let best = null;
    let bestCost = Infinity;
    for (const c of pool) {
      if (chosen.includes(c)) continue;
      const cand = rarityFill([...chosen, c]);
      const n = cost([...table, cand]);
      if (n < bestCost) {
        bestCost = n;
        best = c;
      }
    }
    chosen.push(best);
  }
  const profile = rarityFill(chosen);
  table.push(profile);
  const n = cost(table);
  curve.push(n);
  process.stderr.write(
    `profile ${String(p).padStart(2)}  ${profile.join(' ')}   ` +
      `${n}  (${(n / totalIn).toFixed(5)})\n`,
  );
}

console.log('\n// derived by tools/deriveprofiles.js');
console.log('export const PROFILES = [');
for (const p of table) console.log(`  ${JSON.stringify(p.join(''))},`);
console.log('];');

console.log('\n| profiles | encoded chars | chars per byte | gain |');
console.log('|---|---|---|---|');
for (let i = 0; i < curve.length; i++) {
  const gain = i === 0 ? '' : (((curve[i - 1] - curve[i]) / totalIn) * 100).toFixed(4) + ' %';
  console.log(
    `| ${i + 1} | ${curve[i]} | ${(curve[i] / totalIn).toFixed(5)} | ${gain} |`,
  );
}

File: rust/examples/deriveprofiles.rs
// This Source Code Form is subject to the terms of the Mozilla Public
// License, v. 2.0. If a copy of the MPL was not distributed with this
// file, You can obtain one at https://mozilla.org/MPL/2.0/.

//! Derives the donor profile table of specification Section 8.2.
//!
//! Section 17.5 requires it: the table shipped with 0.4.0 was fitted against
//! the 0.3.0 R-Set, which held `-` and not NUL. The membership changed, `k` can
//! now reach eight, and the eighth rank had never been fitted at all.
//!
//! A profile is an ordered ranking of eight alphabet characters. A segment
//! whose mask has `k` bits set spends the first `k` of them as stand-ins, so
//! only those `k` become unusable as literals inside it -- which is why the
//! table is searched by encoded size and not by character frequency: what
//! matters is how often a donor's own character turns up in the text it would
//! have to stand in for.
//!
//! The search is greedy in two nested ways, as the 0.3.0 derivation was.
//! Profiles are added one at a time, each chosen to help most given the ones
//! already in the table; within a profile, positions are filled left to right,
//! the rest falling back to the rarity order. A table of fewer than
//! `NUM_PROFILES` is evaluated by repeating its last entry: a duplicate profile
//! is never the better of the two, so the encoder behaves as if it were absent.
//!
//!     python3 tools/traincorpus.py
//!     cargo run --release --example deriveprofiles

use std::fs;

use base91z::tables::{tuning, ALPHABET, NUM_PROFILES};

const POOL_SIZE: usize = 20;
const RANKS: usize = 8;

fn main() {
    let dir = std::env::args().nth(1).unwrap_or("bench/train".into());
    let mut paths: Vec<_> = fs::read_dir(&dir)
        .unwrap_or_else(|e| panic!("{dir}: {e} -- run python3 tools/traincorpus.py"))
        .filter_map(|e| e.ok())
        .map(|e| e.path())
        .filter(|p| p.extension().is_some_and(|x| x == "train"))
        .collect();
    paths.sort();
    let files: Vec<Vec<u8>> = paths.iter().map(|p| fs::read(p).unwrap()).collect();
    assert!(!files.is_empty(), "no training data -- run python3 tools/traincorpus.py");
    let total: usize = files.iter().map(|f| f.len()).sum();
    eprintln!("training on {} files, {total} bytes", files.len());

    // Candidate pool: the rarest alphabet characters in the training text.
    // Letters and digits are excluded on principle rather than by frequency --
    // a rare capital is rare when all text is counted together and common in
    // the one file that uses it, so a letter donor breaks segments in bursts.
    // `-` is excluded because Section 8.2 forbids it as a donor.
    let mut counts = [0usize; 256];
    for f in &files {
        for &b in f {
            counts[b as usize] += 1;
        }
    }
    let mut pool: Vec<u8> = ALPHABET
        .iter()
        .copied()
        .filter(|&c| c != b'-' && !c.is_ascii_alphanumeric())
        .collect();
    pool.sort_by_key(|&c| (counts[c as usize], c));
    pool.truncate(POOL_SIZE);
    eprintln!(
        "pool: {}",
        pool.iter().map(|&c| c as char).collect::<String>()
    );

    let rarity_fill = |prefix: &[u8]| -> [u8; RANKS] {
        let mut out = [0u8; RANKS];
        let mut n = 0;
        for &c in prefix.iter().chain(pool.iter()) {
            if n == RANKS {
                break;
            }
            if !out[..n].contains(&c) {
                out[n] = c;
                n += 1;
            }
        }
        assert_eq!(n, RANKS, "pool too small to fill a profile");
        out
    };

    // A table of `built` real profiles, padded to NUM_PROFILES with repeats.
    let cost = |table: &[[u8; RANKS]]| -> usize {
        let mut full = [[0u8; RANKS]; NUM_PROFILES];
        for (i, slot) in full.iter_mut().enumerate() {
            *slot = table[i.min(table.len() - 1)];
        }
        tuning::set_profiles(full);
        files.iter().map(|f| base91z::encode_plain(f).len()).sum()
    };

    let mut table: Vec<[u8; RANKS]> = Vec::new();
    let mut curve: Vec<usize> = Vec::new();
    for p in 0..NUM_PROFILES {
        let mut chosen: Vec<u8> = Vec::new();
        for _ in 0..RANKS {
            let mut best = None;
            let mut best_cost = usize::MAX;
            for &c in &pool {
                if chosen.contains(&c) {
                    continue;
                }
                let mut trial = chosen.clone();
                trial.push(c);
                let mut cand = table.clone();
                cand.push(rarity_fill(&trial));
                let n = cost(&cand);
                if n < best_cost {
                    best_cost = n;
                    best = Some(c);
                }
            }
            chosen.push(best.expect("pool exhausted"));
        }
        table.push(rarity_fill(&chosen));
        let n = cost(&table);
        curve.push(n);
        eprintln!(
            "profile {p}  {}   {n}  ({:.5})",
            table[p].iter().map(|&c| c as char).collect::<String>(),
            n as f64 / total as f64
        );
    }

    println!("\n// derived by examples/deriveprofiles.rs");
    println!("pub const PROFILES: [[u8; 8]; {NUM_PROFILES}] = [");
    for prof in &table {
        let cs: Vec<String> = prof.iter().map(|&c| format!("b'{}'", c as char)).collect();
        println!("    [{}],", cs.join(", "));
    }
    println!("];");

    println!("\n| profiles | 1 | 2 | 3 | 4 |");
    println!("|---|---|---|---|---|");
    print!("| gain |");
    for i in 0..curve.len() {
        if i == 0 {
            print!(" — |");
        } else {
            print!(
                " {:.3} % |",
                (curve[i - 1] as f64 - curve[i] as f64) / total as f64 * 100.0
            );
        }
    }
    println!();
}
